File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_133_house_robber.ts
function rob(nums: number[]): number {
    /**
     * Dynamic Programming approach:
     * - For each house, we choose max of:
     *   1. Rob current house + max money from i-2 houses
     *   2. Skip current house and take max money from i-1 houses
     * 
     * Time: O(n), Space: O(1)
     */
    if (nums.length === 0) return 0;
    if (nums.length === 1) return nums[0];
    
    // prev2: max money robbed up to i-2
    // prev1: max money robbed up to i-1
    let prev2 = nums[0];
    let prev1 = Math.max(nums[0], nums[1]);
    
    for (let i = 2; i < nums.length; i++) {
        const current = Math.max(nums[i] + prev2, prev1);
        prev2 = prev1;
        prev1 = current;
    }
    
    return prev1;
}

// Example usage:
// console.log(rob([1, 2, 3, 1]));      // Output: 4
// console.log(rob([2, 7, 9, 3, 1]));   // Output: 12

File: src/my_project/interviews/top_150_questions_round_22/ex_133_house_robber.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        """
        Dynamic Programming approach:
        - For each house, we choose max of:
          1. Rob current house + max money from i-2 houses
          2. Skip current house and take max money from i-1 houses
        
        Time: O(n), Space: O(1)
        """
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        
        # prev2: max money robbed up to i-2
        # prev1: max money robbed up to i-1
        prev2 = nums[0]
        prev1 = max(nums[0], nums[1])
        
        for i in range(2, len(nums)):
            current = max(nums[i] + prev2, prev1)
            prev2 = prev1
            prev1 = current
        
        return prev1